File: N-ary Tree Preorder Traversal/C#/solution1.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> children;

    public Node(){}
    public Node(int _val,IList<Node> _children) {
        val = _val;
        children = _children;
}
*/
public class Solution {
    public IList<int> Preorder(Node root) {
        IList<int> traversal = new List<int>();
        if(root != null) {
            helper(root, traversal);
        }
        return traversal;
    }
    
    private void helper(Node root, IList<int> list) {

        list.Add(root.val);
        foreach(Node node in root.children) {
            helper(node, list);
        }
    }
}

File: N-ary Tree Preorder Traversal/Python3/solution1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
""" Standard preorder traversal pattern. 
Rather than using left/right we instead itinerate on the children nodes"""
class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        arr = []
        if root == None:
            return arr
        return self.helper(root, arr)
    
    def helper(self, root: 'Node', arr):
        arr.append(root.val)
        
        for child in root.children:
            self.helper(child, arr)
            
        return arr